internal/typeinfo/typeinfo_test.go

```diff
@@ -0,0 +1,161 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReflectSimpleConcurrent(t *testing.T) {
+	type mystruct struct{}
+	var st mystruct
+	wg := sync.WaitGroup{}
+
+	// Set up some concurrent access.
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			_, _ = GetTypeInfo(st)
+			wg.Done()
+		}()
+	}
+
+	info, err := GetTypeInfo(st)
+	assert.Nil(t, err)
+
+	assert.Equal(t, reflect.Struct, info.Type.Kind())
+	assert.Equal(t, "mystruct", info.Type.Name())
+
+	wg.Wait()
+}
+
+func TestReflectStruct(t *testing.T) {
+	type something struct {
+		ID      int64  `db:"id"`
+		Name    string `db:"name,omitempty"`
+		NotInDB string
+	}
+
+	s := something{
+		ID:      99,
+		Name:    "Chainheart Machine",
+		NotInDB: "doesn't matter",
+	}
+
+	info, err := GetTypeInfo(s)
+	assert.Nil(t, err)
+
+	assert.Equal(t, reflect.Struct, info.Type.Kind())
+	assert.Equal(t, "something", info.Type.Name())
+
+	assert.Len(t, info.TagToField, 2)
+
+	id, ok := info.TagToField["id"]
+	assert.True(t, ok)
+	assert.Equal(t, "ID", id.Name)
+	assert.False(t, id.OmitEmpty)
+
+	name, ok := info.TagToField["name"]
+	assert.True(t, ok)
+	assert.Equal(t, "Name", name.Name)
+	assert.True(t, name.OmitEmpty)
+}
+
+func TestReflectNonStructType(t *testing.T) {
+	var i int
+	var s string
+	var mymap map[string]string
+	var myM M
+
+	{
+		info, err := GetTypeInfo(i)
+		assert.Equal(t, fmt.Errorf("can only reflect struct type"), err)
+		assert.Equal(t, &Info{}, info)
+	}
+	{
+		info, err := GetTypeInfo(s)
+		assert.Equal(t, fmt.Errorf("can only reflect struct type"), err)
+		assert.Equal(t, &Info{}, info)
+	}
+	{
+		info, err := GetTypeInfo(mymap)
+		assert.Equal(t, fmt.Errorf("can only reflect struct type"), err)
+		assert.Equal(t, info, &Info{})
+	}
+	{
+		info, err := GetTypeInfo(myM)
+		assert.Equal(t, fmt.Errorf("can only reflect struct type"), err)
+		assert.Equal(t, &Info{}, info)
+	}
+}
+
+func TestReflectBadTagError(t *testing.T) {
+	{
+		type s1 struct {
+			ID int64 `db:"id,bad-juju"`
+		}
+		ss := s1{ID: 99}
+		_, err := GetTypeInfo(ss)
+		assert.Equal(t, fmt.Errorf(`unexpected tag value "bad-juju"`), err)
+	}
+	{
```

Comment:
I think we talked about avoiding these sub-branches in a different review point. Can we please avoid them as they artificially space-out the testing without adding much value?  If these are independent tests, then they shoudl be different functions. If they are the same test, then they can just be normal code, spaced out appropriately.

For the particular case at hand, this sort of test is better organized in a table-styled test. For example, pretty much all of them try to get the type info of some value, and return an error. So you may have something like this:
```
type tagErrorTest struct {
    value any
    error string
}

var tagErrorTests = []tagErrorTest{{
    value: &struct{ ID int64 `db:"id,bad-juju"` } {99}
    error: "error message,
}, {
    ...
}}

func ... {
    for _, test := range tagErrorTests {
    }
}
```

In other words, the repetitive part of the test is extracted out from the key logic, so the latter may be done only once.

These points also apply for the former test case.

Reply:
Changed where applicable.